v2/devdoc/logging.md

```diff
@@ -0,0 +1,221 @@
+# `c-logging` requirements
+
+## Basic functionality
+
+`c-logging` shall expose APIs allowing the user to log events that can be used for observing the software activity, debugging, etc.
+
+The exposed API shall allow creating contexts both dynamically and on the stack and using those contexts for logging.
+
+The API shall allow easy formatting of values to be placed in the contexts by using printf-like formatting.
+
+```
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line_no, const char* format, ...);
+
+#define LOGGER_LOG(log_level, log_context, format, ...) \
+    logger_log(log_level, log_context, __FILE__, __FUNCTION__, __LINE__, format MU_IFCOMMALOGIC(MU_COUNT_ARG(__VA_ARGS__)) __VA_ARGS__)
+
+// macro that can be used to create a dynamically allocated context
+#define LOG_CONTEXT_CREATE(dest_log_context, parent_context, ...) \
+    ...
+
+// destroy a dynamically allocated context
+#define LOG_CONTEXT_DESTROY(log_context_handle) \
+    ...
+
+// macro that can be used to create a context on the stack
+#define LOG_CONTEXT_LOCAL_DEFINE(log_context, parent_context, ...) \
+    ...
+
+// when a user wants to specify properties for a context, they can use LOG_CONTEXT_PROPERTY
+// LOG_CONTEXT_PROPERTY(property_type, property_name, property_value)
+
+// when a user wants to specify string fields in a context using printf-style formatting, they can use LOG_CONTEXT_STRING_PROPERTY
+// LOG_CONTEXT_STRING_PROPERTY(property_name, property_conversion_specifier, property_value)
+```
+
+Example usage:
+
+```c
+    // chained allocated context
+    LOG_CONTEXT_HANDLE context_1;
+    LOG_CONTEXT_CREATE(context_1, NULL, LOG_CONTEXT_STRING_PROPERTY(name, "%s", "haga"));
+    LOG_CONTEXT_HANDLE context_2;
+    LOG_CONTEXT_CREATE(context_2, context_1, LOG_CONTEXT_STRING_PROPERTY(last_name, "%s", "uaga"), LOG_CONTEXT_PROPERTY(int32_t, age, 42));
+
+    LOGGER_LOG(LOG_LEVEL_CRITICAL, context_1, "log context 1 data, value of some_var = %d", 42);
```

Comment:
>context [](http://example.com/codeflow?start=51&length=7)

calling these with "context" was in my mind unexpected, since this data (value of some var) is precisely not part of the context. 

Errr... I don't have a better poetry:(

Maybe just replace context with "handle"?

Reply:
rephrased a bit